Soundboard app

robertwu1 opened the pull request:
New Sample that allows users to continuously play sounds. Please look at the README.md for more info

dturner commented:
This looks really cool. Let's discuss how this could fit with our current samples in our next meeting.

robertwu1 commented on samples/shared/SynthSound.h:
```diff
@@ -0,0 +1,139 @@
+    inline void setAmplitude(float amplitude) {
+        mAmplitudes[0] = amplitude * .2f;
+        mAmplitudes[1] = amplitude;
+        mAmplitudes[2] = amplitude * .1f;
```
@philburk I followed your advice and added multiple overtones. I also added attack and delay and sustain. Can you let me know if this logic makes sense with you?

robertwu1 replied:
Seems like adding a large attack and the following delay makes the sound worse. Reversed the 6th commit "https://github.com/google/oboe/pull/1354/commits/a6224c9b8374a1f6b4018fd0c35986796faf2e30"

robertwu1 replied:
After doing more research, the strong attack followed by the delay makes it like a marimba instead of a piano so reverting makes sense if we wanted a synth sound instead of sounding like a marimba

philburk reviewed (changes requested):
Fantastic! You even gave it a cool icon. This is a great app.

philburk commented on samples/SoundBoard/src/main/java/com/google/oboe/samples/soundboard/MainActivity.java:
```diff
@@ -0,0 +1,133 @@
+        }
+    }
+
+    void createMusicTiles(Context context) {
```
Never mind, you already did.

philburk commented:
I notice that DrumThumper has an average 32 msec tap-to-tone latency and SoundBoard has 45.
That could be due to the bufferSize, touch input technique, or attack time of the sound.
Worth investigating.

philburk commented:
The app sounds a bit quite compared to other apps. Maybe boost the volume a bit, even if maxVoices*amplitude > 1.0.

robertwu1 commented on samples/SoundBoard/local.properties:
```diff
@@ -0,0 +1,8 @@
+## This file must *NOT* be checked into Version Control Systems,
```
Thanks!

robertwu1 commented on samples/SoundBoard/src/main/java/com/google/oboe/samples/soundboard/MusicTileView.java:
```diff
@@ -0,0 +1,143 @@
+    private native void tap(long engineHandle, boolean isDown, int audioSource);
+    private boolean isInitialOnDrawCall = true;
+
+    public MusicTileView(Context context, ArrayList<Rect> rectangles, long engineHandle) {
```
I don't see a grid being much better than what we are doing here: https://stackoverflow.com/questions/14917982/how-to-draw-a-grid-of-rectangle-shaped-objects. It seems like we need to keep track of them anyways and a 2d array might be more of a pain than it is worth

robertwu1 commented:
> I notice that DrumThumper has an average 32 msec tap-to-tone latency and SoundBoard has 45.
> That could be due to the bufferSize, touch input technique, or attack time of the sound.
> Worth investigating.

Thanks! Instead waiting for the onDraw routine, we can call the sound directly!

robertwu1 commented on samples/shared/SynthSound.h:
```diff
@@ -0,0 +1,118 @@
+                for (int j = 0; j < kNumSineWaves; ++j) {
+                    audioData[i] += sinf(mPhases[j]) * mAmplitudes[j] * mAttackCount / kAttackPeriods;
+
+                    mPhases[j].store(mPhases[j] + mPhaseIncrements[j]);
```
Thanks!

robertwu1 commented on samples/shared/SynthSound.h:
```diff
@@ -0,0 +1,118 @@
+            }
+        } else {
+            memset(audioData, 0, sizeof(float) * numFrames);
+            // Remove some noise by letting the current sine wave complete.
```
This is great! Thanks Phil!

robertwu1 commented on samples/SoundBoard/src/main/java/com/google/oboe/samples/soundboard/MusicTileView.java:
```diff
@@ -0,0 +1,143 @@
+
+    private int getIndexFromLocation(PointF pointF) {
+        for (int i = 0; i < mRectangles.size(); i++) {
+            if (pointF.x > mRectangles.get(i).left &&
```
Rectangle.contains() doesn't handle PointF. It doesn't make sense to cast this so resolving

robertwu1 commented on samples/SoundBoard/src/main/cpp/SoundBoardEngine.cpp:
```diff
@@ -0,0 +1,109 @@
+    }
+}
+
+void SoundBoardEngine::tapSource(bool isDown, int32_t audioSource) {
```
Done!

robertwu1 commented:
> The app sounds a bit quite compared to other apps. Maybe boost the volume a bit, even if maxVoices*amplitude > 1.0.

Thanks! I changed the amplitude to .2 from .7. I noticed when I changed the amplitude to 1.0, the sound quality is really bad, even when I lowered the volume. I presume this is from some part of the stack clipping the audio.

robertwu1 commented:
@philburk Thanks for the thorough review! I addressed all the comments. It seems like compared to a tap to tone latency of 34ms with this app and 35ms with DrumThumper, a random top piano app I downloaded had 67ms latency. After considering that the tap latency of the display, it seems like this app is magnitudes faster than what is currently out there. Hopefully by releasing this sample app, we can get current app creators to convert to Oboe.

@dturner I can depreciate MegaDrone in an upcoming PR if we decide to submit this. I realized that everything you want to test/show with MegaDrone is already done with soundboard.

Thanks for all your suggestions and let me know if you have any questions!

philburk commented on samples/SoundBoard/src/main/cpp/SoundBoardEngine.cpp:
```diff
@@ -42,8 +42,12 @@ SoundBoardEngine::~SoundBoardEngine() {
 
-void SoundBoardEngine::tapSource(bool isDown, int32_t audioSource) {
-    mSynth->tapSource(isDown, audioSource);
+void SoundBoardEngine::setNoteOff(int32_t noteIndex) {
```
noteOn() is not normally a "set" method because we are triggering an event, not setting/getting a property.

robertwu1 replied:
Makes sense to me! Done!

philburk commented on samples/SoundBoard/src/main/cpp/Synth.h:
```diff
@@ -54,9 +54,13 @@ class Synth : public IRenderableAudio, public ITappable {
-        mOscs[source].setWaveOn(isOn);
-    };
+    void setNoteOff(int32_t noteIndex) {
+        mOscs[noteIndex].setWaveOn(false);
```
I just realized that you are not using a voice allocator. That may be beyond the scope of this example. Or we can consider it for a later PR.

robertwu1 replied:
What is a voice allocator? If you mean dynamically allocating voices instead of using an array, we can do this in another PR. This one is getting unwieldy

philburk commented on samples/SoundBoard/src/main/java/com/google/oboe/samples/soundboard/MusicTileView.java:
```diff
@@ -134,7 +127,14 @@ public boolean onTouchEvent(MotionEvent event) {
+            // Call c++ native code to set each index add/remove sound
+            for (int i = 0; i < mRectangles.size(); i++) {
+                if (mIsPressedPerRectangle[i]) {
+                    setNoteOn(mEngineHandle, i);
```
This could cause multiple noteOn() events if a finger is held down while another finger is moved.

robertwu1 replied:
Thanks for bringing this up! I have changed the logic when the finger has moved

philburk commented on samples/shared/SynthSound.h:
```diff
@@ -25,22 +25,21 @@ constexpr int32_t kDefaultSampleRate = 48000;
     ~SynthSound() = default;
 
     void setWaveOn(bool isWaveOn) {
         if (!mIsWaveOn && isWaveOn) {
```
This edge detection should occur at the UI level. Then you could use the same UI to trigger MIDI events, which do not use the set/get paradigm.

If a noteOn() occurs while a note is already on then it should retrigger the envelope. This may not be obvious if using a steady sustain value.

robertwu1 replied:
Thanks for giving me a suggested fix below. I have done the following instead:

    void noteOn() {
        mTrigger = true; // start a note envelope
        mAmplitudeScaler = kSustainMultiplier;
    }

    void noteOff() {
        mAmplitudeScaler = kReleaseMultiplier;
    }

philburk commented on samples/shared/SynthSound.h:
```diff
@@ -65,54 +62,42 @@ class SynthSound : public IRenderableAudio {
             memset(audioData, 0, sizeof(float) * numFrames);
             // Remove some noise by letting the current sine wave complete.
             // Since all the frequencies are multiples of mPhases[0], if the first wave complete,
             // so should the others. Use an extra release period to reduce even more noise.
```
This comment does not match the code.

robertwu1 replied:
Nice catch!

philburk commented on samples/shared/SynthSound.h:
```diff
@@ -65,54 +62,42 @@ class SynthSound : public IRenderableAudio {
-        for (int i = 0; i < kNumSineWaves; i++) {
-            mPhaseIncrements[i] = (kTwoPi * mFrequencies[i]) / static_cast<double>(mSampleRate);
-        }
+        mPhaseIncrement = kTwoPi * mFrequency / static_cast<double>(mSampleRate);
```
A divide is surprisingly inefficient compared to a multiply. Cache the inverse and use that in your real-time code. This is particularly important if you a modulating the frequency.

    mSamplePeriod = 1.0 / mSampleRate;

then later

    mPhaseIncrement = ..... * mSamplePeriod;

robertwu1 replied:
mPhaseIncrement is never used in any division so it's fine as is. Thanks for bringing this up though! I'll watch out for optimizations like this in the future

philburk replied:
> mPhaseIncrement is never used in any division so it's fine as is.

But a divide is used to calculate the mPhaseIncrement. It doesn't matter so much here because you are not varying the mPhaseIncrement. But if you add vibrato or portamento then it would be worth making this optimization.

robertwu1 replied:
I added a comment for curious readers

philburk commented on samples/SoundBoard/src/main/java/com/google/oboe/samples/soundboard/MusicTileView.java:
```diff
@@ -0,0 +1,143 @@
+    private SparseArray<PointF> mLocationsOfFingers;
+
+    private long mEngineHandle = 0;
+    private native void setNoteOn(long engineHandle, int noteIndex);
```
Rather than calling a native method directly, call a TileListener interface that has onTileOn(index) and onTimeOff(index) methods. Then you could reuse this for other use cases including MIDI, or even games that do not use notes.  You can implement onTile(index) to call noteOn(index).

robertwu1 replied:
Done!

philburk commented on samples/shared/SynthSound.h:
```diff
@@ -0,0 +1,103 @@
+    void renderAudio(float *audioData, int32_t numFrames) override {
+        if (mIsWaveOn){
+            for (int i = 0; (i < numFrames) && (mMasterAmplitude > kMasterAmplitudeCutOff); ++i) {
+                mMasterAmplitude.store(mMasterAmplitude * kSustainMultiplier);
```
Race condition here. You could use an std::atomic<bool> to trigger the "envelope" atomically.

    noteOn() {
       trigger = true; // start a note envelope
       mIsWaveOn = true;
       mAmplitudeScaler = kDecayScaler;
    }

   noteOff() {
       mAmplitudeScaler = kReleaseScaler;
    }

then from your render thread

    if (trigger.exchange(false)) {
        mMasterAmplitude = 1.0;
    } else {
        mMasterAmplitude *= kWhatever;
    }

So mMasterAmplitude will only be modified from a single thread.

robertwu1 replied:
Thanks!

philburk commented on samples/shared/SynthSound.h:
```diff
@@ -0,0 +1,103 @@
+                    mPhase.store(mPhase - kTwoPi);
+                }
+            }
+        } else {
```
You can eliminate this second block. The only difference from above is the amplitude scaler.

robertwu1 replied:
Good idea! Done!

robertwu1 commented:
@philburk Thanks for submitting more comments. Let me know if you have any more feedback!

philburk commented on samples/SoundBoard/src/main/java/com/google/oboe/samples/soundboard/MusicTileView.java:
```diff
@@ -0,0 +1,147 @@
+package com.google.oboe.samples.soundboard;
+
```
Needs AOSP header.

robertwu1 replied:
Thanks!

philburk commented on samples/SoundBoard/src/main/java/com/google/oboe/samples/soundboard/NoteListener.java:
```diff
@@ -0,0 +1,20 @@
+package com.google.oboe.samples.soundboard;
+
```
Needs AOSP header.

robertwu1 replied:
Thanks!

robertwu1 replied:
Removed this file actually

philburk commented on samples/SoundBoard/src/main/java/com/google/oboe/samples/soundboard/TileListener.java:
```diff
@@ -0,0 +1,7 @@
+package com.google.oboe.samples.soundboard;
+
```
Needs AOSP header. OR define inside MusicTileView.

robertwu1 replied:
Removed file

robertwu1 commented on samples/SoundBoard/src/main/java/com/google/oboe/samples/soundboard/MusicTileView.java:
```diff
@@ -19,19 +20,16 @@
     private boolean[] mIsPressedPerRectangle;
     private Paint mPaint;
     private SparseArray<PointF> mLocationsOfFingers;
+    private NoteListener mNoteListener;
```
I moved the interface and the class inside MusicTileView. Let me know if this looks right

robertwu1 commented on samples/SoundBoard/src/main/java/com/google/oboe/samples/soundboard/MusicTileView.java:
```diff
@@ -75,26 +73,38 @@ public boolean onTouchEvent(MotionEvent event) {
                     }
                 }
+
+                // Now go through the rectangles to see if they have changed
```
Yes, I added some comments to the previous for loop

robertwu1 commented:
@philburk Thanks for adding more comments. They are very helpful for me learning best practices in Java

philburk commented on samples/SoundBoard/src/main/java/com/google/oboe/samples/soundboard/MusicTileView.java:
```diff
@@ -20,16 +36,41 @@
-    private NoteListener mNoteListener;
+    private TileListener mNoteListener;
+
+    public interface TileListener {
```
This goes inside the MusicTileView class as a public inner class.

philburk commented on samples/SoundBoard/src/main/java/com/google/oboe/samples/soundboard/MusicTileView.java:
```diff
@@ -20,16 +36,41 @@
+        public void onTileOff(int index);
+    }
+
+    public class NoteListener implements TileListener {
```
This goes in its own file like before. Sorry that I was not clear.

robertwu1 replied:
Thanks for going over this with me offline. Done!

philburk commented on samples/SoundBoard/src/main/java/com/google/oboe/samples/soundboard/MusicTileView.java:
```diff
@@ -20,16 +36,41 @@
+        }
+    }
+
+    public MusicTileView(Context context, ArrayList<Rect> rectangles, long mEngineHandle) {
```
MusicTileView should not know anything about the synthesizer.

robertwu1 replied:
Thanks!

philburk commented on samples/SoundBoard/src/main/java/com/google/oboe/samples/soundboard/MusicTileView.java:
```diff
@@ -73,6 +114,9 @@ public boolean onTouchEvent(MotionEvent event) {
             // Move each point from it's current point to the new point.
             case MotionEvent.ACTION_MOVE: {
+                // Create an array to check for finger changes as multiple fingers may be on the
+                // same tile. This two-pass algorithm records the difference before changing the
```
"changing the figuring"?

robertwu1 replied:
Nice catch!

philburk commented on samples/shared/SynthSound.h:
```diff
@@ -100,6 +100,9 @@ class SynthSound : public IRenderableAudio {
     std::atomic<float> mFrequency { kDefaultFrequency };
     std::atomic<int32_t> mSampleRate { kDefaultSampleRate };
     void updatePhaseIncrement(){
+        // Note how there is a division here. If this file is changed so that updatePhaseIncrement
```
Good.

robertwu1 commented:
@philburk Addressed your last set of comments. Let me know if I can submit this

philburk reviewed (approved):
Looks great. Thanks for adding a cool new sample app.
